### ISSUE TITLE
Pretest setTiming error

### ISSUE THREAD

**Issue by alphatsai** [psi46/pxar#315]
Hi,

Just wanna to understand if I should be very worry about the error massage during the setTiming in pretest?
As in my experiment, it will be done in the end.
But I just worry if the reason is from the setup or something has to be concerned. 
The error and DEBUG massage like the following: (Using module with TBM08b)
http://cmsdoc.cern.ch/~jtsai/pixel/pxar/commander_Pretest.txt

Many thanks,
Jui-Fa

**Comment by simonspa** [comment]
Hi Jui-Fa,

what do you mean by "it will be done in the end"? The timing test is crucial to get working register settings for the readout - without those all data is garbled and/or missing. If all other tests work, you don't need a timing test because your TBM apparently already has correct register settings.

In general, the answer to your question "should I worry about errors?" is plainly Yes! - otherwise it would not be an error but a "look at me if you have nothing else to look at" message such as INFO or DEBUG.
However, for the timing test things are a bit different: here we play with registers that influence the integrity of the data. We expect to see valid data with all ROCs present etc only of we have proper settings. This means: when scanning register ranges this might lead to corrupt data.
What should always be correct though is the number of expected events as well as the TBM header and trailer (@drberry85 correct me if I'm wrong).

**Comment by drberry85** [comment]
Simon,

Unfortunately, if the timings are wrong then the data can get completely mangled. Particularly if the TBM phase is wrong or the token or header/tailer delays aren't set right. In these cases the data may look unrecognizable. In Jui-Fa case, the TBM phases were not right. He was using 00000000, which destroyed the data integrity.
```
a0d5 c0e3 a040 8000 1f8 21f8 1f8 21f8 5f8 35f8 1f8 2079 7c3 2800 3f 2fff fff 2fff fff 2fff fff 2fff fff 2fff fff 2fff fff 2fff fff 2fff fff
``` 
You can see that the last part of the data is 2ff ffff which should be ffff ffff (system idle), but it looks like 2 bits were lost and a bit was flipped. The timing test then tested a bunch of different ROC delays, which failed, and it then tested the next PLL phase:
```
Testing TBM Phase: 00000100 160 MHz PLL: 0 400MHz PLL: 1
```
Which worked great:
```
[11:58:35.510]    DEBUG: <PixTestPretest.cc/setTimings:L413> Number of Errors: 0
[11:58:35.510]    DEBUG: <PixTestPretest.cc/setTimings:L414> Number of Events: 500
[11:58:35.510]     INFO: Decoding statistics:
[11:58:35.510]     INFO:   General information:
[11:58:35.510]     INFO: 	 16bit words read:         12034
[11:58:35.510]     INFO: 	 empty events:             984
[11:58:35.510]     INFO: 	 valid events with pixels: 16
[11:58:35.510]     INFO: 	 valid pixel hits:         17
[11:58:35.510]     INFO:   Event errors: 	           0
[11:58:35.510]     INFO: 	 start marker:             0
[11:58:35.510]     INFO: 	 stop marker:              0
[11:58:35.510]     INFO: 	 overflow:                 0
[11:58:35.510]     INFO: 	 invalid 5bit words:       0
[11:58:35.510]     INFO: 	 invalid XOR eye diagram:  0
[11:58:35.510]     INFO:   TBM errors: 		           0
[11:58:35.510]     INFO: 	 flawed TBM headers:       0
[11:58:35.511]     INFO: 	 flawed TBM trailers:      0
[11:58:35.511]     INFO: 	 event ID mismatches:      0
[11:58:35.511]     INFO:   ROC errors: 		           0
[11:58:35.511]     INFO: 	 missing ROC header(s):    0
[11:58:35.511]     INFO: 	 misplaced readback start: 0
[11:58:35.511]     INFO:   Pixel decoding errors:	   0
[11:58:35.511]     INFO: 	 pixel data incomplete:    0
[11:58:35.511]     INFO: 	 pixel address:            0
[11:58:35.511]     INFO: 	 pulse height fill bit:    0
[11:58:35.511]     INFO: 	 buffer corruption:        0
[11:58:35.511]     INFO:    ----------------------------------------------------------------------
[11:58:35.511]     INFO:    Good Timings Found!!!
[11:58:35.511]     INFO:    ----------------------------------------------------------------------
[11:58:35.511]     INFO: Setting TBM Phases to 00000100 160 MHz PLL: 0 400MHz PLL: 1
[11:58:35.511]     INFO: Setting ROC Phases to 01100100
```
So the tested worked as it was suppose to. Now the question is should we suppress the deserializer errors when the timings are being set? I would say yes, or we could just tell the user to ignore them. Also, we may want to change the loop order. E.g. check a different TBM phase before changing the ROC phase. I think this should be decided after we get more user feedback.

TLDR: Jui-Fa, your module is fine. It's just how the timing test works right now, but may be changed in the future.

Doug

**Comment by simonspa** [comment]
Hi Doug,

thanks for that piece of information, the bits with the TBM trailer/header I didn't know.

If you really want to suppress the warnings/errors you could temporarily set the verbosity level to QUIET by calling

```
Log::ReportingLevel() = Log::FromString("QUIET");
```

and later back to the original setting. However, I'm not sure we should really do that, let's see what the feedback returns.

**Comment by ursl** [comment]
Hi all,

I would be in favor of suppressing error messages when doing the
timing test (or any other test that uses r/o integrity as a figure of
merit).

Running into errors is a "legitimate" part of this test, but reporting
pages of errors is very misleading for users. Error messages should
stick out and worry the user; if they are part of a normal test
procedure this is not possible and really relevant error messages can
get lost.

We should not suppress error messages for tests or operations when no
error messages are expected, of course.

Cheers,
--U.

**Comment by alphatsai** [comment]
Hi Doug and Simon,

Thanks for your replay.
I can understand why you put setTiming in prestet to correct the phases of TBM now.
But here is another problem comes out after I tested many modules.
One of our modules can be got right TBM phase after prestest::setTiming in -20˚ in pXar v1.8.3.[1]
But when I tried again in this morning with same temperature and the version of pXar, the module seems can not to find good phase during setTiming, so the process was hanging on there and kept looking for.[2]
In the end, I stopped and interrupted it manually.
Here are log files
[1] http://cmsdoc.cern.ch/~jtsai/pixel/pxar/elcomandante_M0284_PretestAndFullTest_Done.txt
[2] http://cmsdoc.cern.ch/~jtsai/pixel/pxar/elcomandante_M0284_Pretest_HangOn.txt

Do you have any suggestions how to deal with this situations?
(BTW, this module was tested successfully in the older version of pXar v1.8.0+67~gdea5405 which was without setTiming in pretest.)

Many thanks,
Jui-Fa

Urs Langenegger <notifications@github.com<mailto:notifications@github.com>> 於 2015年3月20日 上午8:44 寫道：

Hi all,

I would be in favor of suppressing error messages when doing the
timing test (or any other test that uses r/o integrity as a figure of
merit).

Running into errors is a "legitimate" part of this test, but reporting
pages of errors is very misleading for users. Error messages should
stick out and worry the user; if they are part of a normal test
procedure this is not possible and really relevant error messages can
get lost.

We should not suppress error messages for tests or operations when no
error messages are expected, of course.

Cheers,
--U.

**Comment by ursl** [comment]
Hi all,

I just realized that Doug put in the timing test into the doTest
sequence. It is for me not at all clear that is a feature. Unless I
hear a good reason for it, I'll remove it from there on Monday and it
will be available in the pretest for manual execution only.

Cheers,
--U.

**Comment by alphatsai** [comment]
Hi Urs,

Does it mean I can update the pXar and the pretest in head version now is without setTiming?

Thanks,
Jui-Fa


Urs Langenegger <notifications@github.com<mailto:notifications@github.com>> ©ó 2015¦~3¤ë20¤é ¤U¤È2:49 ¼g¹D¡G

Hi all,

I just realized that Doug put in the timing test into the doTest
sequence. It is for me not at all clear that is a feature. Unless I
hear a good reason for it, I'll remove it from there on Monday and it
will be available in the pretest for manual execution only.

Cheers,
--U.

¡X

**Comment by drberry85** [comment]
Jui-Fa,

No, currently it is still there. But I am working on an alternative solution where the timing test only runs if the timings are bad. Which I think is the right thing to do. But Urs can comment further.

Doug

**Comment by ursl** [comment]
Hi Jui-Fa, 

does the new test work on your 'problematic'  module?

The timing test is still in there, but it checks whether something needs to be done. Experience on working and non-working modules would be helpful. I don't know whether a hard removal from the pretest::doTest function is required. 

Cheers,
--U.
